engine/spverrors/unfold.go

```diff
@@ -0,0 +1,92 @@
+package spverrors
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/bitcoin-sv/spv-wallet/models"
+)
+
+type errListUnwrapper interface {
+	Unwrap() []error
+}
+
+// UnfoldError unfolds the error chain into a single string
+// example:
+//
+//	[errType1] outer error message -> [errType2] inner error message -> [errTypeN] innermost error message
+//
+// if error message is contained by outer error, it will be omitted, like:
+//
+//	[errType1] error message -> [errType2]
+//
+// if error is a joined error, it will be unfolded as:
+//
+//	[type] ([errType1] joined error1 AND [errType2] joined error2)
+//
+// for SPVError, it will print the status code in parentheses:
+//
+//	[models.SPVError(404)] error message
+func UnfoldError(err error) string {
+	if err == nil {
+		return ""
+	}
+
+	prevMsg := ""
+	result := strings.Builder{}
+	for current := err; current != nil; current = errors.Unwrap(current) {
+		msg := current.Error()
+
+		if prevMsg != "" {
+			result.WriteString(" -> ")
+		}
+
+		printTypename(current, &result)
+
+		if errListUnwrapper, ok := current.(errListUnwrapper); ok {
+			unfoldJoinedErrors(errListUnwrapper, &result)
+			break // joned errors cannot be unfolded to keep this as chain of errors instead of a tree
+		}
+
+		if prevMsg == "" || !strings.Contains(prevMsg, msg) {
```

Comment:
I'm not sure if this is clear, I'm rather used to have redundant message on each level then suddenly see a lack of message on the root cause of whole stack

Reply:
I thought the same but then I realized that the message can be the same on several (not both), but several levels.

Consider `pkgerrors.Wrap(err, "SOME TEXT")`
The output according to your suggestion would be:
```
[*errors.withStack] ERROR MESSAGE: SOME TEXT -> [*errors.withMessage] -> ERROR MESSAGE: SOME TEXT -> [*actualErrType] SOME TEXT
```

Take into account that there might be several levels of wrapping and the texts are usually longer than here.
My proposition would be:
```
[*errors.withStack] ERROR MESSAGE: SOME TEXT -> [*errors.withMessage] -> [*actualErrType] 
```

Reply:
If we change Wrapf to use error.WithMessage, we can get rid of one level of this nesting.
Beside that, as I said, I (and probably many other developers) used to have duplicated info on many levels of the errors nesting. But at least we don't need to search for the info what was wrong, because it is on the level we search for.